pkg/go/graph/weighted_graph.go

```diff
@@ -0,0 +1,399 @@
+		}
+
+		tupleCyles, err := wg.calculateNodeWeight(node, visited, ancestorPath, tupleCycleDependencies)
+		if err != nil {
+			return err
+		}
+		if len(tupleCyles) > 0 {
+			for _, tupleCycle := range tupleCyles {
+				fmt.Println("Tuple cycle is ", tupleCycle)
+			}
+			return fmt.Errorf("%d tuple cycles found without resolution", len(tupleCyles))
+		}
+	}
+	return nil
+}
+
+func (wg *WeightedAuthorizationModelGraph) calculateNodeWeight(nodeID string, visited map[string]bool, ancestorPath []*WeightedAuthorizationModelEdge, tupleCycleDependencies map[string][]*WeightedAuthorizationModelEdge) ([]string, error) {
+	tupleCycles := make([]string, 0)
+
+	if visited[nodeID] {
+		return nil, nil
+	}
+
+	if wg.nodes[nodeID].nodeType == SpecificType || wg.nodes[nodeID].nodeType == SpecificTypeWildcard {
+		return nil, nil
+	}
+
+	visited[nodeID] = true
+
+	for _, edge := range wg.edges[nodeID] {
+		if len(edge.weights) != 0 {
+			continue
+		}
+		if edge.to.nodeType == SpecificType || edge.to.nodeType == SpecificTypeWildcard {
+			edge.weights = make(map[string]int)
+			uniqueLabel := edge.to.uniqueLabel
+			if edge.to.nodeType == SpecificTypeWildcard {
+				uniqueLabel = uniqueLabel[:len(uniqueLabel)-2]
+			}
+			edge.weights[uniqueLabel] = 1
+			continue
+		}
+
+		tcycle, err := wg.calculateEdgeWeight(edge, ancestorPath, visited, tupleCycleDependencies)
+		if err != nil {
+			tupleCycles = append(tupleCycles, tcycle...)
+			return tupleCycles, err
+		}
+		if len(tcycle) > 0 {
+			tupleCycles = append(tupleCycles, tcycle...) // verify if does not exist first
+		}
+	}
+
+	var err error
+	tupleCycles, err = wg.calculateNodeWeightFromTheEdges(nodeID, tupleCycleDependencies, tupleCycles)
+	if err != nil {
+		return tupleCycles, err
+	}
+	return tupleCycles, nil
+}
+
+func (wg *WeightedAuthorizationModelGraph) calculateEdgeWeight(edge *WeightedAuthorizationModelEdge, ancestorPath []*WeightedAuthorizationModelEdge, visited map[string]bool, tupleCycleDependencies map[string][]*WeightedAuthorizationModelEdge) ([]string, error) {
+	if edge.from.uniqueLabel == edge.to.uniqueLabel {
+		edge.weights = make(map[string]int)
+		edge.weights["R#"+edge.to.uniqueLabel] = Infinite
+		tupleCycleDependencies[edge.to.uniqueLabel] = append(tupleCycleDependencies[edge.to.uniqueLabel], edge)
+		return []string{edge.from.uniqueLabel}, nil
+	}
+
+	ancestorPath = append(ancestorPath, edge)
+	tupleCycle, err := wg.calculateNodeWeight(edge.to.uniqueLabel, visited, ancestorPath, tupleCycleDependencies)
+	if err != nil {
+		return tupleCycle, err
+	}
+
+	if len(edge.to.weights) == 0 {
+		if wg.isTupleCycle(edge.to.uniqueLabel, ancestorPath) {
+			edge.weights = make(map[string]int)
+			edge.weights["R#"+edge.to.uniqueLabel] = Infinite
+			tupleCycleDependencies[edge.to.uniqueLabel] = append(tupleCycleDependencies[edge.to.uniqueLabel], edge)
+			tupleCycle = append(tupleCycle, edge.to.uniqueLabel)
+			return tupleCycle, nil
+		}
+		return tupleCycle, fmt.Errorf("model cycle")
```

Comment:
Create an variable for `fmt.Errorf("model cycle")` which we can then reference/type assert.

Reply:
noted